Add new hardware and software metrics

koute opened the pull request:
This PR adds new hardware/software telemetry to Substrate.

The following extra information about the system is gathered (**Linux-only**):
- CPU name
- CPU core count
- RAM size
- Linux kernel version
- Linux distribution
- Is the node running on a VM?

The following benchmarks are ran on startup (all OSes):
- CPU speed (hashrate of BLAKE2b-256 in MB/s)
- Memory speed (how many MB/s it can `memcpy`)
- Sequential disk write speed
- Random disk write speed

The benchmarks are ran on every startup and in total should take less than ~1s. They are deliberately kept very simple as to not become a maintenance burden.

I've also changed how the node reports its `version`; previously it appended the current CPU ISA + the OS + the environment to the version and sent it as one field (e.g. `0.9.17-75dd6c7d0-x86_64-linux-gnu`) while now that field contains *only* the version (e.g. `0.9.17-75dd6c7d0`) and the rest of the information is transmitted as their own fields.

Fixes (partially) https://github.com/paritytech/substrate/issues/8944

Polkadot PR (should me mergeable independently now, so not marking as companion): https://github.com/paritytech/polkadot/pull/5206

Cumulus PR (should me mergeable independently now, so not marking as companion): https://github.com/paritytech/cumulus/pull/1113

`substrate-telemetry` PR: https://github.com/paritytech/substrate-telemetry/pull/464

cc @emostov @jsdw 

## Questions you might have

### Why is the system information Linux-only?

The majority of nodes are running on Linux, so it makes sense to start there. Besides, gathering this information on Linux is pretty trivial.

### Why not use an external crate to get the system information?

Apparently (or so I've heard) we used such crates in the past and we had problems with them; since gathering this is pretty simple *anyway* I see no point it adding new, potentially janky dependencies.

### Are these benchmarks reliable?

In general from what I can see - yes, however the numbers they produce obviously *do* vary from run-to-run and *can* change depending on whether something else also ran in the background on that machine. But *on average* across the whole network this shouldn't matter. We could periodically rerun them in the future, but for now I propose that we should just add them as-is and see whether any potential noise will actually be a problem or not.

### Is this going to be useful?

I think it will. Although it's hard to tell without actually, you know, adding those metrics in and seeing the results.

### How will those be displayed?

These metrics are printed out in the console (screenshot from our benchmarking machine):

![console](https://user-images.githubusercontent.com/246574/159016865-086c7249-8f96-4d03-acfe-691601f42ae1.png)

And also in `substrate-telemetry` (work in progress; **incomplete**; also, I had only one node connected since I'm only testing it locally, so this is not completely representative of how it will actually look):

![telemetry](https://user-images.githubusercontent.com/246574/159016924-ac0cd481-53b2-419e-87c7-95fb21f7f34d.png)

As you can see there's a new tab/category in the upper right corner on which you can click, and which will bring you to this screen where you'll see a bunch of tables with aggregate statistics for a given chain, showing the most common values for each category. (Somewhat inspired by the Steam Hardware Survey.) This will display all of the information gathered here along with the relative benchmark results as compared to our bechmarking machine. (So we'll see what fraction of the network is running faster/slower hardware.)

wigy-opensource-developer commented on client/service/src/builder.rs:
```diff
@@ -487,8 +491,66 @@ where
+	let sysinfo = crate::sysinfo::gather_sysinfo();
+	let hwbench = crate::sysinfo::gather_hwbench(database_path);
+
+	info!("💻 Operating system: {}", TARGET_OS);
+	info!("💻 CPU architecture: {}", TARGET_ARCH);
+	if !TARGET_ENV.is_empty() {
+		info!("💻 Target environment: {}", TARGET_ENV);
+	}
```
Just looking at these log lines, I would get the impression that the properties of the running system are listed, not those of the build target. I know it is really an edge-case, but foreign ELF formats can be loaded and run emulated on a different system. Are we okay with ignoring those fringe usages?

koute replied:
Hmm... well, that is a good point; I don't think we have to care about this *in general* though since those should mostly be *really* fringe cases, and detecting this will most likely not be easy. (That said, if anyone has any counterpoints here or any good ideas how to handle this in a reasonable way I'm all ears.)

I guess the most likely cases here would be either someone running the Linux binary on a BSD system, or someone running an amd64 binary on an M1 Mac (but we don't provide binaries for macOS, so they'd have to compile it themselves, and if they're compiling it themselves then why not compile a native aarch64 binary in the first place and run that?).

wigy-opensource-developer reviewed (commented):
I understand you did not want to depend on an external crate to gather these data. But are you afraid if you extracted your benchmark code into a separate crate, other people would start asking new features into it unrelated to what Substrate needs?

koute commented:
> I understand you did not want to depend on an external crate to gather these data. But are you afraid if you extracted your benchmark code into a separate crate, other people would start asking new features into it unrelated to what Substrate needs?

Well, I guess we *could* chuck it into a separate crate, but I'm not entirely convinced that it'd be worth it. And, yes, once you get any actual external users they *do* tend to start asking for new features. (: A major point of this implementation is that it's small, simple and narrow in scope. There's a gazillion of other things a general-purpose sysinfo and/or benchmarking crate would have to support *besides* what we support here. (Just compare our ~100 lines of code which gather all the Linux sysinfo we need with the `sysinfo` crate which is 10k lines of code.)

davxy commented on client/service/src/sysinfo.rs:
```diff
@@ -0,0 +1,402 @@
+	hwbench
+}
+
+#[cfg(target_os = "linux")]
+#[test]
+fn test_gather_sysinfo_linux() {
+	let sysinfo = gather_sysinfo();
+	assert!(sysinfo.cpu.unwrap().len() > 0);
+	assert!(sysinfo.core_count.unwrap() > 0);
+	assert!(sysinfo.memory.unwrap() > 0);
+	assert_ne!(sysinfo.is_virtual_machine, None);
+	assert_ne!(sysinfo.linux_kernel, None);
+	assert_ne!(sysinfo.linux_distro, None);
+}
+
+#[test]
+fn test_benchmark_cpu() {
+	assert_ne!(benchmark_cpu(), 0);
+}
+
+#[test]
+fn test_benchmark_memory() {
+	assert_ne!(benchmark_memory(), 0);
+}
+
+#[test]
+fn test_benchmark_disk_sequential_writes() {
+	assert!(benchmark_disk_sequential_writes("./".as_ref()).unwrap() > 0);
+}
+
+#[test]
+fn test_benchmark_disk_random_writes() {
+	assert!(benchmark_disk_random_writes("./".as_ref()).unwrap() > 0);
+}
```
Could you group together the tests under `mod tests`?

davxy commented:
Beside from grouping the tests in their submodule nitpick, LGTM and it is a very cool and useful feature to get some insights about the network nodes

davxy commented on client/service/Cargo.toml:
```diff
@@ -81,6 +82,9 @@ async-trait = "0.1.50"
 directories = "4.0.1"
+regex = "1"
+libc = "0.2"
+nix = "0.23"
```
Looks like `nix` is not used (yet?)

koute replied:
Ah, good catch! I originally used `nix` in an earlier version of this code and forgot to remove it from the `Cargo.toml` after it wasn't needed anymore.

ggwpez commented:
We have a lot of tests that just start a `--dev` node for testing, also in Polkadot.  
See `bin/node/cli/test` in Substrate. Maybe you can add a suppression flag to them as they would otherwise run that over and over.

koute commented:
> We have a lot of tests that just start a `--dev` node for testing, also in Polkadot. See `bin/node/cli/test` in Substrate. Maybe you can add a suppression flag to them as they would otherwise run that over and over.

Considering those benchmarks take less that 1s it shouldn't be too big of a deal in practice, but good point. I've added an extra flag and suppressed those in those tests. (I'll do `polkadot` in a separate PR later since this is not critical and if I don't have to put up a companion I'd rather not.)

koute commented:
Looks like a companion is necessary now anyway since I've added the new CLI argument.

bkchr commented on client/service/src/sysinfo.rs:
```diff
@@ -0,0 +1,407 @@
+// This file is part of Substrate.
```
This should be moved into its own crate.

bkchr commented on client/service/src/config.rs:
```diff
@@ -132,6 +132,14 @@ pub struct Configuration {
 	pub informant_output_format: sc_informant::OutputFormat,
 	/// Maximum number of different runtime versions that can be cached.
 	pub runtime_cache_size: u8,
+	/// Disable automatic hardware benchmarks.
+	///
+	/// By default these benchmarks are automatically ran at startup and measure
+	/// the CPU speed, the memory bandwidth and the disk speed.
+	///
+	/// The results are then printed out in the logs, and also sent as part of
+	/// telemetry, if telemetry is enabled.
+	pub disable_hardware_benchmarks: bool,
```
This should be added by each chain as required, aka we should not put this into the default cli args.

bkchr commented on client/service/src/builder.rs:
```diff
@@ -487,8 +491,72 @@ where
 	)
 	.map_err(|e| Error::Application(Box::new(e)))?;
 
+	let database_path = match config.database {
```
All of this new code here should also be moved into the new crate.

bkchr commented on client/telemetry/src/lib.rs:
```diff
@@ -101,6 +101,54 @@ pub struct ConnectionMessage {
 	/// Node's network ID.
 	pub network_id: String,
+
+	/// Node's OS.
+	pub target_os: String,
+
+	/// Node's ISA.
+	pub target_arch: String,
+
+	/// Node's target platform ABI or libc.
+	pub target_env: String,
+
+	/// Node's software and hardware information.
+	pub sysinfo: Option<SysInfo>,
+
+	/// Node's hardware benchmark results.
+	pub hwbench: Option<HwBench>,
```
The first 3 could stay, but the rest should be removed from the default connection message. This could just be optional messages if a chain is using the benchmarking extension?

bkchr reviewed (changes requested):
I didn't look closely into the benchmarks. However the general structure should be changed. The benchmarks should be optional and chain builders should decide if they want to add it. This means most of the code should be moved into a self contained crate. Then chain builders can call this crate as part of starting the node if they want this (aka we would enable this for Polkadot for example).

koute commented:
Okay, so I've moved the code into a new crate, made `HwBench` into a separate message that goes to telemetry and made it so that the benchmarks have to be manually triggered and registered with the telemetry when building the chain.

As a side effect it's now possible to 1) run the benchmarks very early right before any background threads are started, and 2) rerun the benchmarks at a later time and update the telemetry with the new results (we don't do this currently, but it's now possible).

(Please note: I haven't yet updated the companions in case I'll still have to change something here.)

@bkchr Please let me know if you'd like me to change anything else (e.g. whether you want me to put the non-benchmarking `SysInfo` into its own message; now that I've added support for arbitrary connection messages to telemetry this should be trivial to do.)

bkchr commented on client/telemetry/src/lib.rs:
```diff
@@ -397,6 +504,41 @@ impl TelemetryHandle {
+	/// and update the previously registered connection message.
+	///
+	/// Must be called *after* the telemetry has been initialized.
+	pub fn send_hwbench(&self, hwbench: HwBench) {
```
Why aren't you using the `on_connect_stream`?

I don't really get why you need to modify the telemetry crate?

koute replied:
> I don't really get why you need to modify the telemetry crate?

I don't *need* to per-se; I did it because:

* The new telemetry is effectively part of the protocol since it has special handling on the receiving end of the telemetry, so to me it makes sense to be in the telemetry crate and to make the telemetry crate aware of it.
* The code already has partial support for automatically sending multiple connection messages - the `Node::connection_messages` is a `Vec` so whoever wrote it originally intended for there to be multiple of those, although before before my changes 1) it wasn't actually possible to add new connection messages, and 2) they were sent in the wrong order (but since there was only one connection message it didn't matter).
* It's one less extra task running in the background.

Anyway, I've reverted those changes.

bkchr commented on client/service/src/builder.rs:
```diff
@@ -487,8 +491,36 @@ where
 	)
 	.map_err(|e| Error::Application(Box::new(e)))?;
 
+	let sysinfo = sc_sysinfo::gather_sysinfo();
+
+	info!("💻 Operating system: {}", TARGET_OS);
+	info!("💻 CPU architecture: {}", TARGET_ARCH);
+	if !TARGET_ENV.is_empty() {
+		info!("💻 Target environment: {}", TARGET_ENV);
+	}
```
Could we not also just do all this inside the sysbench crate?

koute replied:
Good idea. Done.

bkchr commented on client/sysinfo/src/sysinfo.rs:
```diff
@@ -0,0 +1,404 @@
+impl Deref for TemporaryFile {
+	type Target = File;
+	fn deref(&self) -> &Self::Target {
+		self.fp.as_ref().unwrap()
+	}
+}
+
+impl DerefMut for TemporaryFile {
+	fn deref_mut(&mut self) -> &mut Self::Target {
+		self.fp.as_mut().unwrap()
```
Please use expect here

bkchr commented on client/sysinfo/src/sysinfo.rs:
```diff
@@ -0,0 +1,404 @@
+		}
+	}
+
+	positions.shuffle(&mut rng());
```
Don't you want to use some fixed seed here to always have this reproducible?

bkchr replied:
It probably doesn't make such a big difference.

koute replied:
Unless I'm missing something I am using a fixed seed? (:

```rust
fn rng() -> rand_pcg::Pcg64 {
	rand_pcg::Pcg64::new(0xcafef00dd15ea5e5, 0xa02bdbf7bb3c0a7ac28fa16a64abf96)
}
```

bkchr replied:
Ohh fuck :D I did not realize that `rng` was defined locally. I did not say anything :P

bkchr commented on client/sysinfo/src/sysinfo.rs:
```diff
@@ -0,0 +1,404 @@
+	};
+
+	// We only wrote half of the bytes hence `SIZE / 2`.
+	crate::sysinfo::benchmark(
```
```suggestion
	benchmark(
```

bkchr commented on client/sysinfo/src/sysinfo.rs:
```diff
@@ -0,0 +1,404 @@
+		Ok(())
+	};
+
+	crate::sysinfo::benchmark(
```
```suggestion
	benchmark(
```

bkchr commented on client/sysinfo/src/lib.rs:
```diff
@@ -0,0 +1,98 @@
+			_ => unreachable!("the `HwBench` always serializes into a JSON object; qed"),
+		};
+		payload.insert("msg".into(), "sysinfo.hwbench".into());
+		while let Some(_) = connect_stream.next().await {
```
```suggestion
		while connect_stream.next().await.is_some() {
```

bkchr commented on client/telemetry/src/lib.rs:
```diff
@@ -101,6 +101,51 @@ pub struct ConnectionMessage {
+	pub is_virtual_machine: Option<bool>,
+}
+
+/// Hardware benchmark results for the node.
+#[derive(Debug, Serialize)]
+pub struct HwBench {
+	/// The CPU speed, as measured in how many MB/s it can hash using the BLAKE2b-256 hash.
+	pub cpu_hashrate_score: u64,
+	/// Memory bandwidth in MB/s, calculated by measuring the throughput of `memcpy`.
+	pub memory_memcpy_score: u64,
+	/// Sequential disk write speed in MB/s.
+	pub disk_sequential_write_score: Option<u64>,
+	/// Random disk write speed in MB/s.
+	pub disk_random_write_score: Option<u64>,
 }
```
I would move this to the sysinfo crate as well.

bkchr reviewed (approved):
Mainly some nitpicks, otherwise it looks good. However, I also don't checked every benchmark into each detail.

arkpar commented:
Minor nit: for the CLI we follow the GNU convention of using `--no-whatever` rather than`--disable-whatever`

koute commented:
I've randomly recompiled and retested this code and.... the hwbench telemetry stopped working. The message was not being sent for some reason, even though I swear it used to work.

It took a while to debug, but it turns out the connection notification stream in telemetry was totally broken - it was using `send` which is *async* and returns a future which needs to be `await`ed on, but the code deliberately ignoring it with a `let _ =`. I have no idea why it worked before. Anyway, I've fixed it to use the synchronous `try_send` instead, and made it so that it won't leak disconnected notifiers while I'm at it.

This should be good to go I think; I'll let it marinate until tomorrow and if there won't be any more comments I'll merge it in.

koute commented:
> Minor nit: for the CLI we follow the GNU convention of using `--no-whatever` rather than`--disable-whatever`

We have the `--disable-log-color` argument, so not entirely it seems... (:

But I don't really care either way; I can change it.

bkchr commented:
> it was using `send` which is _async_ and returns a future which needs to be `await`ed on, but the code was deliberately ignoring it with a `let _ =`. I have no idea why it worked before. Anyway, I've fixed it to use the synchronous `try_send` instead, and made it so that it won't leak disconnected notifiers while I'm at it. (Alas, `retain_mut`'s still unstable, so it's a little ugly.)

Okay that is really bad and your solution is also not really great. I already see people complaining about the warning appearing very often. To the point why it worked before, `send` is internally first taking the same path as `try_send` and if that fails it returns the future to try again. The correct solution here would be to collect these futures and to poll them until they are finished.

bkchr commented:
(If you convert your `warn` into a `debug` you can merge as is and open an issue to implement this properly)

koute commented:
> To the point why it worked before, `send` is internally first taking the same path as `try_send` and if that fails it returns the future to try again. The correct solution here would be to collect these futures and to poll them until they are finished.

Wait, is it? From what I can see `send` doesn't actually do anything besides creating a new struct. ([[1]](https://github.com/rust-lang/futures-rs/blob/77d10e9f1293ab65664c9bc75ed8da4f0d7d8336/futures-util/src/sink/mod.rs#L220), [[2]](https://github.com/rust-lang/futures-rs/blob/77d10e9f1293ab65664c9bc75ed8da4f0d7d8336/futures-util/src/sink/send.rs#L20), [[3]](https://github.com/rust-lang/futures-rs/blob/77d10e9f1293ab65664c9bc75ed8da4f0d7d8336/futures-util/src/sink/feed.rs#L20)) I was convinced it used to work, but this definitely should have not worked. Maybe I'm just going crazy. (Or I'm looking at the wrong thing?)

> and your solution is also not really great. I already see people complaining about the warning appearing very often. [..] The correct solution here would be to collect these futures and to poll them until they are finished.

Unless I'm missing something here AFAIK it shouldn't appear at all in normal circumstances. The `try_send` should always succeed as long as there's space in the channel (that is: as long as the receiver is not stuck; if I'm reading [this code](https://github.com/rust-lang/futures-rs/blob/77d10e9f1293ab65664c9bc75ed8da4f0d7d8336/futures-channel/src/mpsc/mod.rs#L516) right) and this is triggered only when we succeeded (re)connecting to the telemetry. So for this warning to trigger this needs to happen:

1) We get disconnected from the telemetry.
2) We get reconnected to the telemetry.
3) The connection notifier receiver must have not yet processed a notification from **before** the (1) happened.

We're probably probably not going to be disconnected from the telemetry very often (and that prints a warning on its own *anyway*), and on top of that for this warning to trigger the receiver either isn't handling the notifications at all, or is handling them really slow (slower than the time it takes to get disconnected from the telemetry and reconnected again), so I'd argue the `warn!` here is warranted since if it's printed out there's probably a bug somewhere in the code (because it hasn't yet processed the notification from *before* when we got disconnected).

Also, here's a quick test program to make sure this works as I think it works:

```rust
use futures::StreamExt;

#[tokio::main]
async fn main() {
    let (mut tx, mut rx) = futures::channel::mpsc::channel(0);
    tokio::task::spawn_blocking(move || {
        loop {
            tx.try_send(()).unwrap();
            std::thread::sleep(std::time::Duration::from_millis(1));
        }
    });

    loop {
        rx.next().await.unwrap();
        println!("Got message");
    }
}
```

This prints out:

```
Got message
Got message
Got message
Got message
...
```

So `try_send` always succeeds as long as the receiver is processing the messages fast enough, so I think in practice this warning should never actually be triggered. (And if it does I'd like to know, since either something's seriously broken, or this whole thing works entirely different than how I think it works.)

bkchr commented:
> Wait, is it? From what I can see `send` doesn't actually do anything besides creating a new struct. ([[1]](https://github.com/rust-lang/futures-rs/blob/77d10e9f1293ab65664c9bc75ed8da4f0d7d8336/futures-util/src/sink/mod.rs#L220), [[2]](https://github.com/rust-lang/futures-rs/blob/77d10e9f1293ab65664c9bc75ed8da4f0d7d8336/futures-util/src/sink/send.rs#L20), [[3]](https://github.com/rust-lang/futures-rs/blob/77d10e9f1293ab65664c9bc75ed8da4f0d7d8336/futures-util/src/sink/feed.rs#L20)) I was convinced it used to work, but this definitely should have not worked. Maybe I'm just going crazy. (Or I'm looking at the wrong thing?)

No you are right. I didn't checked the code yesterday and was under the impression that the model was right in my brain. Sorry! No after thinking again about it this doesn't make any sense, because futures need to be polled to do something...


> So `try_send` always succeeds as long as the receiver is processing the messages fast enough, so I think in practice this warning should never actually be triggered. (And if it does I'd like to know, since either something's seriously broken, or this whole thing works entirely different than how I think it works.)

The point here is you don't know what is on the other side and why it is maybe not processing messages fast enough. I would argue that if you for example reconnected very fast and the other side still doesn't has processed the first reconnect, there is no harm in just dropping any further reconnect message, because they will process the waiting message at some point. We can also not consume too much memory, because there is only room for one element in the channel.

TDLR: Either remove the log completely or turn it into a debug please.

koute commented:
Okay, I've changed it to a `debug` log and also renamed the command-line argument as @arkpar requested. Should be good now; I'll also update the companions.

koute commented:
bot merge

koute commented:
Merged in manually since the bot was still somehow watching the companion PR (which shouldn't have to be a companion now and should be mergable independently, so no point in making it harder than it needs to be).
